plugin/keyring_vault/vault_curl.cc

```diff
@@ -0,0 +1,305 @@
+#include <my_global.h>
+#include <boost/scope_exit.hpp>
+#include <algorithm>
+#include "my_rdtsc.h"
+#include "sql_error.h"
+#include "mysqld.h"
+#include "mysql/service_thd_wait.h"
+#include "vault_curl.h"
+#include "vault_memory.h"
+#include "vault_base64.h"
+
+namespace keyring
+{
+
+static const size_t max_response_size = 32000000;
+static const long timeout = 300; // 5m timeout
+static MY_TIMER_INFO curl_timer_info;
+static ulonglong last_ping_time;
+static bool was_thd_wait_started = false;
+static const ulonglong slow_connection_threshold = 100; // [ms]
+
+class Thd_wait_end_guard
+{
+  public:
+    ~Thd_wait_end_guard()
+    {
+      DBUG_EXECUTE_IF("vault_network_lag", {
+                    was_thd_wait_started = false;   
+        });
+      DBUG_ASSERT(!was_thd_wait_started);
+      if (was_thd_wait_started)
+      {
+        // This should never be called as thd_wait_end should be called at the end of CURL I/O operation.
+        // However, in production the call to thd_wait_end cannot be missed. Thus we limit our trust
+        // to CURL lib and make sure thd_wait_end was called.
+        thd_wait_end(current_thd);
+        was_thd_wait_started = false;
+      }
+    }
+};
+
+static size_t write_response_memory(void *contents, size_t size, size_t nmemb, void *userp)
+{
+  size_t realsize = size * nmemb;
+  if (size != 0 && realsize/size != nmemb)
+    return 0; // overflow
+  Secure_ostringstream *read_data = static_cast<Secure_ostringstream*>(userp);
+  size_t ss_pos = read_data->tellp();
+  read_data->seekp(0, std::ios::end);
+  size_t number_of_read_bytes = read_data->tellp();
+  read_data->seekp(ss_pos);
+
+  if (number_of_read_bytes + realsize > max_response_size)
+    return 0; // response size limit exceeded
+
+  read_data->write(static_cast<char*>(contents), realsize);
+  if (!read_data->good())
+    return 0;
+  return realsize;
+}
+
+int progress_callback(void *clientp, double dltotal, double dlnow, double ultotal, double ulnow)
+//int progress_callback(void *clientp, curl_off_t dltotal, curl_off_t dlnow, curl_off_t ultotal, curl_off_t ulnow)
+{
+  ulonglong curr_ping_time = my_timer_milliseconds();
+
+  DBUG_EXECUTE_IF("vault_network_lag", {
+                curr_ping_time = last_ping_time + slow_connection_threshold + 10; 
+                dltotal = 1;
+                dlnow = 0;
+    });
+
+  BOOST_SCOPE_EXIT(&curr_ping_time, &last_ping_time)
+  {
+    last_ping_time = curr_ping_time;
+  } BOOST_SCOPE_EXIT_END
+  if (!was_thd_wait_started)
+  { 
+    if ((dlnow < dltotal || ulnow < ultotal) && last_ping_time - curr_ping_time > slow_connection_threshold)
+    { 
+      // there is a good chance that connection is slow, thus we can let know the threadpool that there is time
```

Comment:
I have enabled threadpool in MTR test keyring_vault_thd_wait.test. However this is causing a crash. I have filled in a bug PS-244 and disabled calling threadpool's functions from keyring_vault plugin. This should be re-enabled once bug gets fixed.